File: Estudos/3 prova/beecrownd/Array Hash 1257/Main.java
import java.util.*;

public class Main {

    static class Hash {
        int[] tabela;
        int padrao = -1;
        int tam;
        int soma; 

        public Hash(int x) {
            this.tam = x;
            this.tabela = new int[x];
            for (int i = 0; i < x; i++) {
                tabela[i] = padrao;
            }
            this.soma = 0;
        }

        private int h(int elemento) {
            return elemento % tam;
        }

        public void inserir(int elemento) {
            int pos = h(elemento);
            tabela[pos] = elemento; 
            soma += elemento;       
        }

        public int getSoma() {
            return soma;
        }
    }

    public static int obterValor(int linha, String str) {
        int valor = 0;
        for (int i = 0; i < str.length(); i++) {     
            int c = str.charAt(i) - 'A';             
            valor += linha + c + i;                  
        }
        return valor;
    }

    public static void main(String[] args) {
        Scanner ent = new Scanner(System.in);

        int casos = ent.nextInt();

        while (casos > 0) {
            int qtdLinhas = ent.nextInt();
            ent.nextLine();

            Hash hs = new Hash(qtdLinhas);

            for (int i = 0; i < qtdLinhas; i++) {
                String entrada = ent.nextLine().trim();
                int valor = obterValor(i, entrada);
                hs.inserir(valor);
            }

            System.out.println(hs.getSoma());

            casos--;
        }

        ent.close();
    }
}

File: Estudos/3 prova/beecrownd/Array Hash 1257/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *tabela;
    int tam;
    int padrao;
    int soma;
} Hash;

void inicializarHash(Hash *h, int tamanho) {
    int i;
    h->tam = tamanho;
    h->padrao = -1;
    h->soma = 0;
    h->tabela = (int*) malloc(tamanho * sizeof(int));
    for (i = 0; i < tamanho; i++) {
        h->tabela[i] = h->padrao;
    }
}

int funcaoHash(Hash *h, int elemento) {
    if (h->tam == 0) return 0;
    return elemento % h->tam;
}

void inserirHash(Hash *h, int elemento) {
    int pos = funcaoHash(h, elemento);
    h->tabela[pos] = elemento;  
    h->soma += elemento;        
}

int obterValor(int linha, char texto[]) {
    int i = 0;
    int valor = 0;
    while (texto[i] != '\0') {
        int posLetra = texto[i] - 'A';  
        valor += linha + posLetra + i;  
        i++;
    }
    return valor;
}

int main() {
    int casos;
    int i;

    if (scanf("%d", &casos) != 1) {
        return 0;
    }

    while (casos > 0) {
        int qtdLinhas;
        scanf("%d", &qtdLinhas);

        Hash h;
        inicializarHash(&h, qtdLinhas);

        for (i = 0; i < qtdLinhas; i++) {
            char texto[60];
            scanf("%s", texto);  

            int valor = obterValor(i, texto);
            inserirHash(&h, valor);
        }

        printf("%d\n", h.soma);

        free(h.tabela);
        casos--;
    }

    return 0;
}
